Feature/r20 41

rafRangler opened the pull request:
# Description

This PR addresses the changes to move from figma v2 to figma v3 -> allows for dynamic function matching to process tokens. This change requires a figmaFileConfig.json file to read multiple urls vs a single url.

*Changes still need to be added to convert the nodes collection to design tokens. 
*Work with designers to finalize naming convention for figma designs

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B
- [ ] 

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

rafRangler committed:
feat(radius-styles.ts): added a method getBlobs() that resolves api promises. Will be refactored

rafRangler committed:
feat(radius-styles.ts figmaparser.utils.ts): updated figmaResolver

This code change uses the new way of resolving parser functions dynamically. This allows us to add
new methods for validating and parsing each file vs tweaking our token processing methods whenever a
format, prefixes, or naming of the figma file changes.

rafRangler committed:
refactor(figaresolver.utils.ts): refactor

transformNodes() method moved to figmaResolver. transformNodes() is the primary function that
handles mapping of functions for tokenGroups

rafRangler commented on packages/cli/src/commands/styles/lib/radius-styles.ts:
```diff
@@ -35,14 +49,16 @@ export const generateGlobalStyles = async ({
-  const tokenGroups = await loadFile({ url, token: userToken }).then(getTokens).then(groupByType);
+  const figmaNodes = await getFigmaBlobs(userToken).then((data) => transformNodes(data));
+  //return all tokens;
+  console.log(figmaNodes);
```
I can remove line 55. I think 53 should be commented out for now until we've tested the new parsing handlers.

rafRangler commented on packages/cli/src/commands/styles/utils/figma.utils.ts:
```diff
@@ -22,7 +24,7 @@ export type FigmaFileParams = {
 type FigmaFileNodes = {
   nodes: {
-    [key: string]: NodeRoot,
+    [key: string]: NodeRoot | any ,
```
One of the issues with our existing code is that its taking the type `GetFileNodesResult`, the default return type from the figma api and casting it as any in `figma.utils.ts`. I'll remove the union, but we should look into switching our custom types to use the types provided by `FigmaAPI`
```
    const figmaNode: any = node;
    const root: NodeRoot = extractFirstNode(figmaNode) as NodeRoot;
```
^quick fix but our existing code in `Figma.utils.ts` is casting the `GetFileNodesResult` to any
```
export const getTokens = (data: any) => {
   Promise.resolve(data)
      .then((x) => {
```

Bohdan-Anderson-Rangle replied:
OK I'll add "switch to Figma api types" as a ticket to Jira.

rafRangler commented on packages/cli/src/commands/styles/utils/figmaResolver.utils.ts:
```diff
@@ -0,0 +1,119 @@
+// loader configuration
+// validator 
+// different values from different applications... no frontify great
+//
```
whoops left a comment there. Since its separated by validators and extractors, I'll remove this

rafRangler commented on packages/cli/src/commands/styles/utils/validators/figmaValidators.ts:
```diff
@@ -0,0 +1,27 @@
+  return Object.entries(generateStyleMap(tree.componentSets, filterByTypography)).length > 0 ? true : false;
+};
+  
+export const isTypographyFormat2: FigmaTokenAnalyzer = (tree: NodeRoot) => {
```
For futureproofing this, we want to have separate files for each validator and extractor. `isTypographyFormat2` and `isTypographyFormat1` are different validators. The format below is what we want.

File directory
/validators
  validator1.ts - sanofi
  validator2.ts - nike
  validator3.ts - new balance

/extractors 
  extractor1.ts - sanofi
  extractor2.ts - nike
  extractor3.ts - new balance

validatorReader.ts 
extractorReader.ts

rafRangler committed:
feat(figma.utils, figmaresolver.ts): pR fixes

This PR includes the new FigmaResolver.utils.ts file to handle different ways to validate and parse
figma data

rafRangler committed:
fix(radius-styles.ts): linting errors
